From a16c4b6bb0b126047c68dafbdf6311ac1586fc0b Mon Sep 17 00:00:00 2001
From: Bryan Call <bcall@apache.org>
Date: Tue, 4 Mar 2025 11:39:44 -0800
Subject: [PATCH] Fix chunked pipelined requests (#12071)

* Fix pipelined request for chunked bodies

This addresses a variety of bugs concerning pipelined requests. In
particular, the HttpTunnel logic had fundamentally assumed that it could
consume all bytes available in the producer's reader. If a request was
pipelined after a previous request that had a chunked body, this would
result in the second request being unparsed and either sent along to the
origin or dropped on the floor, depening on configuration. This adds an
explicit autest for pipelined requests and addresses these issues.

This patch largely does the following:

1. Updates the copy_partial_post_data data to take the number of bytes
   it consumes rather than consuming all bytes in the reader. It also
   now returns the number of bytes it consumes, which the tunnel needs
   to keep track of the number of bytes it processes.
2. Previous to this patch, the HttpTunnel assumed that it could consume
   all bytes in the reader originally passed to it (all bytes in
   init_bytes_done). This simply will not work for pipelined requests.
   This addresses this issue by adding a new variable to the tunnel:
   bytes_consumed. This way the tunnel can keep track of how many bytes
   it consumed while processing the request body, which allows the
   HttpSM to later process just the right number of bytes from its
   reader rather than eating into any pipelined requests that follow it.
3. The HttpSM must not consume bytes from its client reader that are
   pipelined requests. It now uses the tunnel's processing
   bytes_consumed to process bytes from its reader rather than simply
   consuming all read_available() bytes from it.

* Fix bytes consumed chunk computation

Fix a possible miscalculation of bytes consumed while parsing chunked
content.

* is_read_closed fix

* Verify expected responses are received.

* Updated formatting

---------

Co-authored-by: Brian Neradt <brian.neradt@gmail.com>
---
 proxy/ProxyTransaction.cc                     |   6 +
 proxy/ProxyTransaction.h                      |   1 +
 proxy/http/HttpSM.cc                          |  70 ++++--
 proxy/http/HttpSM.h                           |  14 +-
 proxy/http/HttpTunnel.cc                      | 171 +++++++++------
 proxy/http/HttpTunnel.h                       | 151 +++++++++++--
 proxy/http2/Http2Stream.cc                    |   6 +
 proxy/http2/Http2Stream.h                     |   1 +
 src/traffic_server/FetchSM.cc                 |   3 +-
 tests/gold_tests/pipeline/pipeline.test.py    | 137 ++++++++++++
 tests/gold_tests/pipeline/pipeline_client.py  | 104 +++++++++
 tests/gold_tests/pipeline/pipeline_server.py  | 201 ++++++++++++++++++
 .../gold_tests/redirect/redirect_post.test.py |   1 +
 .../timeout/inactive_client_timeout.test.py   |   9 +-
 tests/gold_tests/timeout/slow_server.yaml     |  37 +++-
 15 files changed, 796 insertions(+), 116 deletions(-)
 create mode 100644 tests/gold_tests/pipeline/pipeline.test.py
 create mode 100644 tests/gold_tests/pipeline/pipeline_client.py
 create mode 100644 tests/gold_tests/pipeline/pipeline_server.py

diff --git a/proxy/http/HttpTunnel.h b/proxy/http/HttpTunnel.h
index 3ce994b11ed..3aac38aca68 100644
--- a/proxy/http/HttpTunnel.h
+++ b/proxy/http/HttpTunnel.h
@@ -145,14 +152,54 @@ struct ChunkedHandler {
   /// If @a size is zero it is set to @c DEFAULT_MAX_CHUNK_SIZE.
   void set_max_chunk_size(int64_t size);
 
-  // Returns true if complete, false otherwise
-  bool process_chunked_content();
-  bool generate_chunked_content();
+  /** Consumes and processes chunked content.
+   *
+   * This consumes data from @a chunked_reader and, if dechunking, writes the
+   * dechunked body to @a dechunked_buffer.
+   *
+   * @return The number of bytes consumed from the chunked reader and true if
+   * the entire chunked content is processed, false otherwise.
+   */
+  std::pair<int64_t, bool> process_chunked_content();
+
+  /** Writes chunked content.
+   *
+   * This reads from @a dechunked_reader and writes chunked formatted content to
+   * @a chunked_buffer.
+   *
+   * @return The number of bytes consumed from the dechunked reader and true if
+   * the entire chunked content is written, false otherwise.
+   */
+  std::pair<int64_t, bool> generate_chunked_content();
 
 private:
-  void read_size();
-  void read_chunk();
-  void read_trailer();
+  /** Read a chunk header containing the size of the chunk.
+   *
+   * @return The number of bytes consumed from the chunked buffer reader.
+   */
+  int64_t read_size();
+
+  /** Read a chunk body.
+   *
+   * This is called after read_size so that the chunk size is known.
+   *
+   * @return The number of bytes consumed from the chunked buffer reader.
+   */
+  int64_t read_chunk();
+
+  /** Read a chunk trailer.
+   *
+   * @return The number of bytes consumed from the chunked buffer reader.
+   */
+  int64_t read_trailer();
+
+  /** Transfer body bytes from the chunked reader.
+   *
+   * This will either simply consume the chunked body bytes in the case of
+   * chunked passthrough, or transfer the chunked body to the dechunked buffer.
+   *
+   * @return The number of bytes consumed from the chunked buffer reader.
+   */
   int64_t transfer_bytes();
 
   constexpr static std::string_view FINAL_CRLF = "\r\n";
